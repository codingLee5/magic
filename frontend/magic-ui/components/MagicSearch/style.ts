import { createStyles } from "antd-style"

export const useStyles = createStyles(({ isDarkMode, token }) => {
	return {
		search: {
			paddingLeft: 6,
			paddingRight: 6,
			borderRadius: 8,
			border: `1px solid transparent`,
			color: isDarkMode ? token.magicColorScales.grey[4] : token.magicColorUsages.text[3],
			backgroundColor: token.magicColorScales.grey[1],
			transition: "0.1s linear width",

			[`&:hover`]: {
				border: `1px solid ${token.colorBorder}`,
				backgroundColor: `${token.magicColorUsages.fill[0]}`,
				color: `${
					isDarkMode ? token.magicColorScales.grey[4] : token.magicColorUsages.text[2]
				}`,
			},
			[`&:active`]: {
				backgroundColor: `${token.magicColorUsages.fill[2]}`,
			},
		},
	}
})
