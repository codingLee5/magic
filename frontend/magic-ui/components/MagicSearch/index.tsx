import { IconSearch } from "@tabler/icons-react"
import type { InputProps, InputRef } from "antd"
import { Input } from "antd"
import { cx } from "antd-style"
import { forwardRef, useState, memo, useEffect } from "react"
import type { CompositionEvent, ChangeEvent } from "react"
import MagicIcon from "../MagicIcon"
import { useStyles } from "./style"

export type MagicSearchProps = InputProps

const MagicSearch = memo(
	forwardRef<InputRef, MagicSearchProps>(({ className, onChange, value, ...props }, ref) => {
		const { styles } = useStyles()

		const [rawValue, setRawValue] = useState(value)
		const [isComposing, setIsComposing] = useState(false)

		useEffect(() => {
			setRawValue(value)
		}, [value])

		return (
			<Input
				ref={ref}
				value={rawValue}
				onCompositionStart={() => setIsComposing(true)}
				onCompositionEnd={(e: CompositionEvent<HTMLInputElement>) => {
					setIsComposing(false)
					// 在输入法结束时触发一次onChange
					onChange?.(e as unknown as ChangeEvent<HTMLInputElement>)
				}}
				onChange={(e) => {
					// 只在非输入法编辑状态下触发
					setRawValue(e.target.value)
					if (!isComposing) {
						onChange?.(e)
					}
				}}
				className={cx(styles.search, className)}
				prefix={<MagicIcon component={IconSearch} size={20} color="currentColor" />}
				{...props}
			/>
		)
	}),
)

export default MagicSearch
