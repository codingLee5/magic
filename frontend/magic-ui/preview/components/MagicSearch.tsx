import React from "react"
import { Space } from "antd"
import MagicSearch from "../../components/MagicSearch"
import ComponentDemo from "./Container"

const SearchDemo: React.FC = () => {
	return (
		<div>
			<ComponentDemo
				title="基础搜索"
				description="最基本的搜索组件"
				code="<MagicSearch placeholder='请输入搜索内容' />"
			>
				<Space>
					<MagicSearch placeholder="请输入搜索内容" style={{ width: 300 }} />
				</Space>
			</ComponentDemo>

			<ComponentDemo
				title="不同尺寸"
				description="支持不同尺寸的搜索框"
				code="size: 'large' | 'middle' | 'small'"
			>
				<Space direction="vertical">
					<MagicSearch placeholder="大尺寸搜索框" size="large" style={{ width: 300 }} />
					<MagicSearch placeholder="默认尺寸搜索框" style={{ width: 300 }} />
					<MagicSearch placeholder="小尺寸搜索框" size="small" style={{ width: 300 }} />
				</Space>
			</ComponentDemo>

			<ComponentDemo
				title="事件处理"
				description="监听搜索和输入事件"
				code="onSearch | onChange"
			>
				<Space direction="vertical">
					<MagicSearch
						placeholder="输入时触发onChange"
						style={{ width: 300 }}
						onChange={(e) => console.log("输入内容:", e.target.value)}
					/>
				</Space>
			</ComponentDemo>
		</div>
	)
}

export default SearchDemo
