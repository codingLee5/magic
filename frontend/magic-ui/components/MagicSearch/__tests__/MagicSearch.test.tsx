import { render, screen } from "@testing-library/react"
import MagicThemeProvider from "../../ThemeProvider"
import { describe, it, expect } from "vitest"
import MagicSearch from "../index"

const renderWithTheme = (component: React.ReactElement) =>
	render(<MagicThemeProvider theme="light">{component}</MagicThemeProvider>)

describe("MagicSearch", () => {
	it("应该正常渲染", () => {
		renderWithTheme(<MagicSearch placeholder="搜索..." />)
		expect(screen.getByPlaceholderText("搜索...")).toBeInTheDocument()
	})

	// 快照测试
	describe("快照测试", () => {
		it("基础搜索框快照", () => {
			const { asFragment } = renderWithTheme(<MagicSearch placeholder="搜索..." />)
			expect(asFragment()).toMatchSnapshot()
		})

		it("带默认值搜索框快照", () => {
			const { asFragment } = renderWithTheme(<MagicSearch defaultValue="默认搜索词" />)
			expect(asFragment()).toMatchSnapshot()
		})

		it("带前缀图标搜索框快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicSearch placeholder="搜索..." prefix={<span>🔍</span>} />,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("带后缀图标搜索框快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicSearch placeholder="搜索..." suffix={<span>📝</span>} />,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("禁用状态搜索框快照", () => {
			const { asFragment } = renderWithTheme(<MagicSearch placeholder="搜索..." disabled />)
			expect(asFragment()).toMatchSnapshot()
		})

		it("大尺寸搜索框快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicSearch placeholder="搜索..." size="large" />,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("小尺寸搜索框快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicSearch placeholder="搜索..." size="small" />,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("带边框搜索框快照", () => {
			const { asFragment } = renderWithTheme(<MagicSearch placeholder="搜索..." bordered />)
			expect(asFragment()).toMatchSnapshot()
		})
	})
})
