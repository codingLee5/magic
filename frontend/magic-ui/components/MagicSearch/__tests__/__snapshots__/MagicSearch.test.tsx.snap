// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MagicSearch > 快照测试 > 基础搜索框快照 1`] = `
<DocumentFragment>
  <span
    class="magic-input-affix-wrapper css-dev-only-do-not-override-yotw4n magic-input-outlined acss-9dldd5 css-var-r1 magic-input-css-var"
  >
    <span
      class="magic-input-prefix"
    >
      <svg
        class="tabler-icon tabler-icon-search "
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 10m-7 0a7 7 0 1 0 14 0a7 7 0 1 0 -14 0"
        />
        <path
          d="M21 21l-6 -6"
        />
      </svg>
    </span>
    <input
      class="magic-input css-dev-only-do-not-override-yotw4n"
      placeholder="搜索..."
      type="text"
      value=""
    />
  </span>
</DocumentFragment>
`;

exports[`MagicSearch > 快照测试 > 大尺寸搜索框快照 1`] = `
<DocumentFragment>
  <span
    class="magic-input-affix-wrapper magic-input-affix-wrapper-lg css-dev-only-do-not-override-yotw4n magic-input-outlined acss-9dldd5 css-var-r6 magic-input-css-var"
  >
    <span
      class="magic-input-prefix"
    >
      <svg
        class="tabler-icon tabler-icon-search "
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 10m-7 0a7 7 0 1 0 14 0a7 7 0 1 0 -14 0"
        />
        <path
          d="M21 21l-6 -6"
        />
      </svg>
    </span>
    <input
      class="magic-input magic-input-lg css-dev-only-do-not-override-yotw4n"
      placeholder="搜索..."
      type="text"
      value=""
    />
  </span>
</DocumentFragment>
`;

exports[`MagicSearch > 快照测试 > 小尺寸搜索框快照 1`] = `
<DocumentFragment>
  <span
    class="magic-input-affix-wrapper magic-input-affix-wrapper-sm css-dev-only-do-not-override-yotw4n magic-input-outlined acss-9dldd5 css-var-r7 magic-input-css-var"
  >
    <span
      class="magic-input-prefix"
    >
      <svg
        class="tabler-icon tabler-icon-search "
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 10m-7 0a7 7 0 1 0 14 0a7 7 0 1 0 -14 0"
        />
        <path
          d="M21 21l-6 -6"
        />
      </svg>
    </span>
    <input
      class="magic-input magic-input-sm css-dev-only-do-not-override-yotw4n"
      placeholder="搜索..."
      type="text"
      value=""
    />
  </span>
</DocumentFragment>
`;

exports[`MagicSearch > 快照测试 > 带前缀图标搜索框快照 1`] = `
<DocumentFragment>
  <span
    class="magic-input-affix-wrapper css-dev-only-do-not-override-yotw4n magic-input-outlined acss-9dldd5 css-var-r3 magic-input-css-var"
  >
    <span
      class="magic-input-prefix"
    >
      <span>
        🔍
      </span>
    </span>
    <input
      class="magic-input css-dev-only-do-not-override-yotw4n"
      placeholder="搜索..."
      type="text"
      value=""
    />
  </span>
</DocumentFragment>
`;

exports[`MagicSearch > 快照测试 > 带后缀图标搜索框快照 1`] = `
<DocumentFragment>
  <span
    class="magic-input-affix-wrapper css-dev-only-do-not-override-yotw4n magic-input-outlined acss-9dldd5 css-var-r4 magic-input-css-var"
  >
    <span
      class="magic-input-prefix"
    >
      <svg
        class="tabler-icon tabler-icon-search "
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 10m-7 0a7 7 0 1 0 14 0a7 7 0 1 0 -14 0"
        />
        <path
          d="M21 21l-6 -6"
        />
      </svg>
    </span>
    <input
      class="magic-input css-dev-only-do-not-override-yotw4n"
      placeholder="搜索..."
      type="text"
      value=""
    />
    <span
      class="magic-input-suffix"
    >
      <span>
        📝
      </span>
    </span>
  </span>
</DocumentFragment>
`;

exports[`MagicSearch > 快照测试 > 带边框搜索框快照 1`] = `
<DocumentFragment>
  <span
    class="magic-input-affix-wrapper css-dev-only-do-not-override-yotw4n magic-input-outlined acss-9dldd5 css-var-r8 magic-input-css-var"
  >
    <span
      class="magic-input-prefix"
    >
      <svg
        class="tabler-icon tabler-icon-search "
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 10m-7 0a7 7 0 1 0 14 0a7 7 0 1 0 -14 0"
        />
        <path
          d="M21 21l-6 -6"
        />
      </svg>
    </span>
    <input
      class="magic-input css-dev-only-do-not-override-yotw4n"
      placeholder="搜索..."
      type="text"
      value=""
    />
  </span>
</DocumentFragment>
`;

exports[`MagicSearch > 快照测试 > 带默认值搜索框快照 1`] = `
<DocumentFragment>
  <span
    class="magic-input-affix-wrapper css-dev-only-do-not-override-yotw4n magic-input-outlined acss-9dldd5 css-var-r2 magic-input-css-var"
  >
    <span
      class="magic-input-prefix"
    >
      <svg
        class="tabler-icon tabler-icon-search "
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 10m-7 0a7 7 0 1 0 14 0a7 7 0 1 0 -14 0"
        />
        <path
          d="M21 21l-6 -6"
        />
      </svg>
    </span>
    <input
      class="magic-input css-dev-only-do-not-override-yotw4n"
      type="text"
      value="默认搜索词"
    />
  </span>
</DocumentFragment>
`;

exports[`MagicSearch > 快照测试 > 禁用状态搜索框快照 1`] = `
<DocumentFragment>
  <span
    class="magic-input-affix-wrapper magic-input-disabled magic-input-affix-wrapper-disabled css-dev-only-do-not-override-yotw4n magic-input-outlined acss-9dldd5 css-var-r5 magic-input-css-var"
  >
    <span
      class="magic-input-prefix"
    >
      <svg
        class="tabler-icon tabler-icon-search "
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 10m-7 0a7 7 0 1 0 14 0a7 7 0 1 0 -14 0"
        />
        <path
          d="M21 21l-6 -6"
        />
      </svg>
    </span>
    <input
      class="magic-input magic-input-disabled css-dev-only-do-not-override-yotw4n"
      disabled=""
      placeholder="搜索..."
      type="text"
      value=""
    />
  </span>
</DocumentFragment>
`;
